File: dart/lib/src/models/created_webhook.dart
//
//  rerout
//  created_webhook.dart
//
//  Official Dart SDK for the Rerout branded-link API.
//
//  Created by Ngonidzashe Mangudya on 20/05/2026.
//  Copyright (c) 2026 Codecraft Solutions
//  Licensed under the MIT License
//  https://codecraftsolutions.co.za
//

import 'package:meta/meta.dart';
import 'package:rerout/src/models/webhook.dart';

/// The result of creating a webhook endpoint.
///
/// The [signingSecret] (`whsec_…`) is returned **once** — store it now to
/// verify inbound deliveries; it is never shown again.
@immutable
class CreatedWebhook {
  /// Creates a [CreatedWebhook].
  const CreatedWebhook({required this.endpoint, required this.signingSecret});

  /// Parses a [CreatedWebhook] from the API JSON envelope.
  factory CreatedWebhook.fromJson(Map<String, dynamic> json) => CreatedWebhook(
    endpoint: Webhook.fromJson(json['endpoint'] as Map<String, dynamic>),
    signingSecret: json['signing_secret'] as String,
  );

  /// The newly-created webhook endpoint.
  final Webhook endpoint;

  /// The signing secret (`whsec_…`). Returned once — persist it now.
  final String signingSecret;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CreatedWebhook &&
          endpoint == other.endpoint &&
          signingSecret == other.signingSecret;

  @override
  int get hashCode => Object.hash(endpoint, signingSecret);

  @override
  String toString() => 'CreatedWebhook(endpoint: $endpoint)';
}

File: php/src/Models/CreatedWebhook.php
<?php

declare(strict_types=1);

namespace Rerout\Models;

use InvalidArgumentException;

/**
 * Result of creating a webhook endpoint.
 *
 * The `signing_secret` (`whsec_…`) is returned **once** — store it now to
 * verify deliveries; it is never shown again.
 */
final readonly class CreatedWebhook
{
    public function __construct(
        public Webhook $endpoint,
        public string $signingSecret,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $endpoint = $data['endpoint'] ?? null;
        if (!is_array($endpoint)) {
            throw new InvalidArgumentException("Expected object field 'endpoint' in CreatedWebhook payload.");
        }

        $secret = $data['signing_secret'] ?? null;
        if (!is_string($secret)) {
            throw new InvalidArgumentException("Expected string field 'signing_secret' in CreatedWebhook payload.");
        }

        /** @var array<string, mixed> $endpoint */

        return new self(
            endpoint: Webhook::fromArray($endpoint),
            signingSecret: $secret,
        );
    }
}
